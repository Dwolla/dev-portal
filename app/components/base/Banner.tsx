import styled from "@emotion/styled";
import classnames from "../../modules/classnames";
import Button from "./Button";
import { PURPLE_DARK, WHITE_PRIMARY, GREY_4 } from "../colors";
import { POPPINS, ROBOTO } from "../typography";
import { breakDown } from "../breakpoints";

// Styled Banner component
const StyledBanner = styled.div`
  background-color: ${PURPLE_DARK};
  padding: 40px;
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
  width: 100%;

  &.overlapped {
    padding-bottom: 75px;
  }

  &.filters {
    padding-bottom: 70px;
  }
`;

const BannerContentWrap = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  > *:last-child {
    margin-bottom: 0;
  }
`;

const StyledIcon = styled.img`
  height: 46px;
  width: 42px;
  margin-bottom: 20px;
`;

const StyledTopic = styled.div`
  max-width: 749px;
  color: ${WHITE_PRIMARY};
  font-family: ${POPPINS};
  font-size: 42px;
  font-weight: 200;
  line-height: 52px;
  text-align: center;
  margin-bottom: 20px;

  @media (${breakDown("sm")}) {
    font-size: 30px;
    line-height: 46px;
  }

  @media (${breakDown("xs")}) {
    font-size: 22px;
    line-height: 33px;
  }
`;

const StyledDescription = styled.div`
  color: ${GREY_4};
  font-family: ${ROBOTO};
  font-size: 19px;
  font-weight: 300;
  line-height: 26px;
  text-align: center;
  margin-bottom: 30px;

  @media (${breakDown("sm")}) {
    font-size: 16px;
    line-height: 20px;
  }
`;

const StyledFilter = styled.div`
  text-transform: uppercase;
  position: absolute;
  bottom: 0;
`;

// Button props
export type ButtonProps = {
  text: string;
  link: {
    href: string;
    external: boolean;
  };
};

// Prop Types
type Props = {
  icon?: string;
  topic: string;
  description: string;
  button?: ButtonProps;
  filterTabs?: JSX.Element;
  variant?: "default" | "overlapped";
};

// Banner component
function Banner({
  icon,
  topic,
  description,
  button,
  filterTabs,
  variant = "default",
}: Props) {
  return (
    <StyledBanner className={classnames({ filters: filterTabs }, variant)}>
      <BannerContentWrap>
        {icon && <StyledIcon src={icon} alt="" />}
        <StyledTopic>{topic}</StyledTopic>
        <StyledDescription>{description}</StyledDescription>
        {button && <Button {...button} size="large" variant="primary" />}
      </BannerContentWrap>
      {filterTabs && <StyledFilter>{filterTabs}</StyledFilter>}
    </StyledBanner>
  );
}

export default Banner;
