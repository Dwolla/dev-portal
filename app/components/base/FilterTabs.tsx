import React from "react";
import styled from "@emotion/styled";
import Select from "./select/Select";
import {
  GREY_1,
  GREY_2,
  GREY_4,
  GREY_6,
  GREY_8,
  WHITE_PRIMARY,
  ORANGE_PRIMARY,
} from "../colors";
import { POPPINS, ROBOTO } from "../typography";
import { breakUp, breakDown } from "../breakpoints";
import classnames from "../../modules/classnames";
import { slugify } from "../../modules/helpers";

// Styles
const TabsContainerStyle = styled.ul`
  margin: unset;
  padding: unset;
  list-style-type: none;
  justify-content: center;
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 30px;

  @media (${breakDown("xs")}) {
    display: none;
  }

  &.sidebar {
    justify-content: left;
    display: block;
  }
`;

export const TabStyle = styled.li`
  height: 25px;
  color: ${GREY_4};
  font-family: ${POPPINS};
  font-size: 13px;
  font-weight: 500;
  line-height: 20px;
  justify-self: center;
  padding-bottom: 30px;
  cursor: pointer;

  :hover {
    color: ${GREY_2};
  }

  :focus {
    outline: none;
    color: ${WHITE_PRIMARY};
  }

  &.is-active {
    color: ${WHITE_PRIMARY};
    border-bottom: 3px solid ${ORANGE_PRIMARY};
  }

  &.sidebar {
    padding: 9px 0 11px 17px;
    height: 38px;
    min-width: 147px;
    color: ${GREY_6};
    background-color: ${WHITE_PRIMARY};
    font-family: ${ROBOTO};
    font-size: 14px;
    font-weight: unset;
    letter-spacing: 0;
    line-height: 19px;
    justify-self: left;
    box-sizing: border-box;
    border: 1px solid ${GREY_2};
    border-bottom: none;

    :last-child {
      border-bottom: 1px solid ${GREY_2};
    }

    :focus,
    :hover {
      color: ${GREY_8};
      padding-left: 16px;
      border-left: 2px solid ${GREY_6};
      background-color: ${GREY_1};
    }

    &.is-active {
      color: ${GREY_8};
      border-left: 2px solid ${ORANGE_PRIMARY};
      padding-left: 16px;
    }
  }
`;

// Tabs are replaced with Select dropdown on mobile
const SelectWrapper = styled.div`
  padding-bottom: 10px;

  @media (${breakUp("sm")}) {
    display: none;
  }

  &.sidebar {
    display: none;
  }
`;

type Tab = string | { value: string; label: string };
type Props = {
  tabs: Array<Tab>;
  filter: Tab | null;
  setFilter: Function;
  variant?: "default" | "sidebar";
};

export default function FilterTabs({
  tabs,
  filter,
  setFilter,
  variant,
}: Props) {
  const formattedTabs = tabs.map((tab) =>
    typeof tab === "string" ? { value: slugify(tab), label: tab } : tab
  );

  const formattedFilter =
    typeof filter === "string" || filter === null
      ? formattedTabs.find((tab) => tab.value === filter)
      : filter;

  function onKeydown(e, tab) {
    if (e.key === "Enter") setFilter(tab);
  }

  return (
    <>
      <TabsContainerStyle className={variant}>
        {formattedTabs.map((tab) => (
          <TabStyle
            tabIndex={0}
            className={classnames(variant, {
              "is-active":
                JSON.stringify(formattedFilter) === JSON.stringify(tab),
            })}
            key={tab.value}
            onClick={() => {
              setFilter(tab);
            }}
            onKeyDown={(keyPress) => onKeydown(keyPress, tab)}
          >
            {tab.label}
          </TabStyle>
        ))}
      </TabsContainerStyle>

      <SelectWrapper className={variant}>
        <Select
          options={formattedTabs}
          selectedValue={formattedFilter}
          setSelectedValue={setFilter}
          autoWidth
        />
      </SelectWrapper>
    </>
  );
}
