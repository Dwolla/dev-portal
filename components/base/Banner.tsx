import styled from "@emotion/styled";
import classnames from "../../modules/classnames";
import Button from "./Button";
import { PURPLE_DARK, WHITE_PRIMARY, GREY_4 } from "../colors";
import { POPPINS, ROBOTO } from "../typography";
import { breakDown } from "../breakpoints";

// Styled Banner component
const StyledBanner = styled.div`
  height: 100%;
  background-color: ${PURPLE_DARK};
  padding: 40px;
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  align-items: center;
  position: relative;
  width: 100%;
  &.filters {
    padding-bottom: 0px;
  }
`;

const StyledIcon = styled.img`
  height: 46px;
  width: 42px;
`;

const StyledTopic = styled.div`
  max-width: 749px;
  color: ${WHITE_PRIMARY};
  font-family: ${POPPINS};
  font-size: 42px;
  font-weight: 200;
  line-height: 52px;
  text-align: center;
  @media (${breakDown("sm")}) {
    font-size: 30px;
    line-height: 46px;
  }
  @media (${breakDown("xs")}) {
    font-size: 22px;
    line-height: 33px;
  }
`;

const StyledDescription = styled.div`
  color: ${GREY_4};
  font-family: ${ROBOTO};
  font-size: 19px;
  font-weight: 300;
  line-height: 26px;
  text-align: center;
  &.filters {
    margin-bottom: 52px;
  }
  @media (${breakDown("sm")}) {
    font-size: 16px;
    line-height: 20px;
  }
`;

const StyledFilter = styled.div`
  text-transform: uppercase;
  margin-top: 52px;
  margin-bottom: 0px;
  position: absolute;
  bottom: 0;
`;

// Button props
export type ButtonProps = {
  text: string;
  link: {
    href: string;
    external: boolean;
  };
};

// Prop Types
type Props = {
  icon?: string;
  topic: string;
  description: string;
  button?: ButtonProps;
  filterTabs?: JSX.Element;
};

// Banner component
function Banner({ icon, topic, description, button, filterTabs }: Props) {
  return (
    <StyledBanner className={classnames({ filters: filterTabs })}>
      {icon && <StyledIcon src={icon} alt="" />}
      <StyledTopic>{topic}</StyledTopic>
      <StyledDescription className={classnames({ filters: filterTabs })}>
        {description}
      </StyledDescription>
      {button && <Button {...button} size="large" variant="primary" />}
      {filterTabs && <StyledFilter>{filterTabs}</StyledFilter>}
    </StyledBanner>
  );
}

export default Banner;
