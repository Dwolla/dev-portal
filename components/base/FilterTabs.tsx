import React from "react";
import Dropdown from "react-dropdown";
import styled from "@emotion/styled";
import arrowDown from "../../public/images/arrow-down-icon.svg";
import arrowUp from "../../public/images/arrow-up-icon.svg";
import { GREY_4, WHITE_PRIMARY, PURPLE_DARK, ORANGE_PRIMARY } from "../colors";
import { POPPINS } from "../typography";
import { minWidth, maxWidth, BREAKPOINT_MOBILE } from "../breakpoints";
import classnames from "../../modules/classnames";

type Props = {
  tabs: string[];
  filter: string;
  setFilter: Function;
};

export default function FilterTabs({ tabs, filter, setFilter }: Props) {
  const dropdownOptions = [
    { value: "ALL", label: "ALL", className: "option" },
    ...tabs.map((item) => ({
      value: item,
      label: item,
      className: "option",
    })),
  ];
  const defaultOption = filter || dropdownOptions[0];

  return (
    <div>
      <TabsContainerStyle>
        <TabStyle
          className={classnames("tab", { "is-active": filter === null })}
          onClick={() => setFilter(null)}
        >
          ALL
        </TabStyle>

        {tabs.map((tab) => (
          <TabStyle
            className={classnames("tab", { "is-active": filter === tab })}
            key={tab}
            onClick={() => setFilter(tab)}
          >
            {tab}
          </TabStyle>
        ))}
      </TabsContainerStyle>

      <Drop>
        <Dropdown
          controlClassName="customControlClass"
          menuClassName="customMenuClass"
          placeholderClassName="customPlaceHolderClass"
          arrowOpen={<span className="arrowOpenClass" />}
          arrowClosed={<span className="arrowClosedClass" />}
          options={dropdownOptions}
          value={defaultOption}
          onChange={(e) => setFilter(e.label)}
        />
      </Drop>
    </div>
  );
}

// Styles
const TabsContainerStyle = styled.div`
  justify-content: center;
  display: grid;
  grid-auto-flow: column;
  grid-column-gap: 30px;
  @media (${maxWidth(BREAKPOINT_MOBILE)}) {
    display: none;
  }
`;

export const TabStyle = styled.div`
  height: 25px;
  color: ${GREY_4};
  font-family: ${POPPINS};
  font-size: 13px;
  font-weight: 500;
  line-height: 20px;
  justify-self: center;
  padding-bottom: 30px;
  cursor: pointer;

  :hover {
    color: ${WHITE_PRIMARY};
  }

  &.is-active {
    color: ${WHITE_PRIMARY};
    border-bottom: 3px solid ${ORANGE_PRIMARY};
  }
`;

const Drop = styled.div`
  padding: 10px;
  @media (${minWidth(BREAKPOINT_MOBILE)}) {
    display: none;
  }

  /* Custom styles that are passed as props to react-dropdown component */
  .customControlClass {
    display: flex;
    justify-content: space-between;
  }
  .customControlClass,
  .customMenuClass {
    background-color: ${PURPLE_DARK};
    color: ${GREY_4};
    font-family: ${POPPINS};
    font-size: 13px;
    font-weight: 500;
    line-height: 20px;
    cursor: pointer;
  }

  .customPlaceHolderClass {
    color: ${WHITE_PRIMARY};
  }

  .option {
    color: inherit;
    margin-top: 5px;
    margin-bottom: 5px;
  }

  .option:hover {
    background-color: inherit;
    color: ${WHITE_PRIMARY};
  }

  .option.is-selected {
    background-color: inherit;
    color: ${WHITE_PRIMARY};
  }

  .arrowOpenClass,
  .arrowClosedClass {
    display: block;
    content: " ";
    padding: 10px;
    width: auto;
  }

  .arrowOpenClass {
    background: url(${arrowUp}) no-repeat center;
  }

  .arrowClosedClass {
    background: url(${arrowDown}) no-repeat center;
  }
`;
