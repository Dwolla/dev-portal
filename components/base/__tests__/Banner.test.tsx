import renderer from "react-test-renderer";
import Banner from "../Banner";
import FilterTabs, { TabStyle } from "../FilterTabs";
import guideIcon from "../../assets/images/content-images/content-icons/guides-icon.svg";

test("Default Banner", () => {
  const tree = renderer
    .create(
      <Banner
        icon={guideIcon}
        topic="Guides"
        description="Step-by-step instructions to get you set up in the Dwolla API"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("Banner - no icon", () => {
  const tree = renderer
    .create(
      <Banner
        topic="Comprehensive API documentation for developers and businesses."
        description="Step-by-step instructions to get you set up in the Dwolla API"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("Banner - with Button", () => {
  const tree = renderer
    .create(
      <Banner
        topic="Comprehensive API documentation for developers and businesses."
        description="Step-by-step instructions to get you set up in the Dwolla API"
        button={{
          text: "Discover the Possibilities",
          link: {
            href: "https://accounts-sandbox.dwolla.com/login",
            external: true,
          },
        }}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test("Banner - with Extra Padding for Overlapping Cards", () => {
  const tree = renderer
    .create(
      <Banner
        topic="Comprehensive API documentation for developers and businesses."
        description="Step-by-step instructions to get you set up in the Dwolla API"
        button={{
          text: "Discover the Possibilities",
          link: {
            href: "https://accounts-sandbox.dwolla.com/login",
            external: true,
          },
        }}
        variant="overlapped"
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

const TABS = [
  { value: "all", label: "ALL" },
  { value: "get-building", label: "GET BUILDING" },
  { value: "prerequisites", label: "PREREQUISITES" },
  { value: "features", label: "FEATURES" },
];

test("Banner - with Filters", () => {
  const setState = jest.fn();
  const tree = renderer.create(
    <Banner
      icon={guideIcon}
      topic="Guides"
      description="Step-by-step instructions to get you set up in the Dwolla API"
      filterTabs={
        <FilterTabs tabs={TABS} filter={TABS[0]} setFilter={setState} />
      }
    />
  );
  tree.root.findAllByType(TabStyle).map((t) => t.props.onClick());

  expect(setState.mock.calls.map((c) => c)).toEqual([...TABS].map((t) => [t]));
  expect(tree.toJSON()).toMatchSnapshot();
});
