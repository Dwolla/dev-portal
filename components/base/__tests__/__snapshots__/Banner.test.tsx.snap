// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner - no icon 1`] = `
<div
  className="default css-qc8b1t-StyledBanner eyaavg40"
>
  <div
    className="css-2ux9ul-BannerContentWrap eyaavg41"
  >
    <div
      className="css-izwvz-StyledTopic eyaavg43"
    >
      Comprehensive API documentation for developers and businesses.
    </div>
    <div
      className="css-wbadgb-StyledDescription eyaavg44"
    >
      Step-by-step instructions to get you set up in the Dwolla API
    </div>
  </div>
</div>
`;

exports[`Banner - with Button 1`] = `
<div
  className="default css-qc8b1t-StyledBanner eyaavg40"
>
  <div
    className="css-2ux9ul-BannerContentWrap eyaavg41"
  >
    <div
      className="css-izwvz-StyledTopic eyaavg43"
    >
      Comprehensive API documentation for developers and businesses.
    </div>
    <div
      className="css-wbadgb-StyledDescription eyaavg44"
    >
      Step-by-step instructions to get you set up in the Dwolla API
    </div>
    <a
      className="primary large css-qnxw47-StyledLink-buttonStyles-buttonStyles e12ysvgk1"
      href="https://accounts-sandbox.dwolla.com/login"
      onClick={[Function]}
      onMouseEnter={[Function]}
      target="_blank"
    >
      Discover the Possibilities
    </a>
  </div>
</div>
`;

exports[`Banner - with Extra Padding for Overlapping Cards 1`] = `
<div
  className="overlapped css-qc8b1t-StyledBanner eyaavg40"
>
  <div
    className="css-2ux9ul-BannerContentWrap eyaavg41"
  >
    <div
      className="css-izwvz-StyledTopic eyaavg43"
    >
      Comprehensive API documentation for developers and businesses.
    </div>
    <div
      className="css-wbadgb-StyledDescription eyaavg44"
    >
      Step-by-step instructions to get you set up in the Dwolla API
    </div>
    <a
      className="primary large css-qnxw47-StyledLink-buttonStyles-buttonStyles e12ysvgk1"
      href="https://accounts-sandbox.dwolla.com/login"
      onClick={[Function]}
      onMouseEnter={[Function]}
      target="_blank"
    >
      Discover the Possibilities
    </a>
  </div>
</div>
`;

exports[`Banner - with Filters 1`] = `
<div
  className="filters default css-qc8b1t-StyledBanner eyaavg40"
>
  <div
    className="css-2ux9ul-BannerContentWrap eyaavg41"
  >
    <img
      alt=""
      className="css-dt0gsi-StyledIcon eyaavg42"
      src="test-image-stub"
    />
    <div
      className="css-izwvz-StyledTopic eyaavg43"
    >
      Guides
    </div>
    <div
      className="css-wbadgb-StyledDescription eyaavg44"
    >
      Step-by-step instructions to get you set up in the Dwolla API
    </div>
  </div>
  <div
    className="css-1wmc1h7-StyledFilter eyaavg45"
  >
    <div>
      <div
        className="css-18c9go6-TabsContainerStyle efq5uj30"
      >
        <div
          className="tab is-active css-rdl1v4-TabStyle efq5uj31"
          onClick={[Function]}
        >
          ALL
        </div>
        <div
          className="tab css-rdl1v4-TabStyle efq5uj31"
          onClick={[Function]}
        >
          GET BUILDING
        </div>
        <div
          className="tab css-rdl1v4-TabStyle efq5uj31"
          onClick={[Function]}
        >
          PREREQUISITES
        </div>
        <div
          className="tab css-rdl1v4-TabStyle efq5uj31"
          onClick={[Function]}
        >
          FEATURES
        </div>
      </div>
      <div
        className="css-n6mzzn-SelectWrapper efq5uj32"
      >
        <div
          className=" css-2b097c-container"
          onKeyDown={[Function]}
        >
          <div
            className=" css-1xxipdn-control"
            onMouseDown={[Function]}
            onTouchEnd={[Function]}
          >
            <div
              className=" css-86rg1c-ValueContainer"
            >
              <div
                className=" css-1me8bqy-singleValue"
              >
                ALL
              </div>
              <input
                aria-autocomplete="list"
                className="css-62g3xt-dummyInput"
                disabled={false}
                id="react-select-2-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                tabIndex="0"
                value=""
              />
            </div>
            <div
              className=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                className=" css-43ykx9-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                className=" css-unjupu-indicatorContainer"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <img
                  alt=""
                  className=" css-vfym8x-StyledDefaultDropdownIcon e1oqaitj0"
                  src="test-image-stub"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Default Banner 1`] = `
<div
  className="default css-qc8b1t-StyledBanner eyaavg40"
>
  <div
    className="css-2ux9ul-BannerContentWrap eyaavg41"
  >
    <img
      alt=""
      className="css-dt0gsi-StyledIcon eyaavg42"
      src="test-image-stub"
    />
    <div
      className="css-izwvz-StyledTopic eyaavg43"
    >
      Guides
    </div>
    <div
      className="css-wbadgb-StyledDescription eyaavg44"
    >
      Step-by-step instructions to get you set up in the Dwolla API
    </div>
  </div>
</div>
`;
